src/index/sparse/sparse_inverted_index.h

```diff
@@ -0,0 +1,457 @@
+#include "knowhere/log.h"
+#include "knowhere/sparse_utils.h"
+#include "knowhere/utils.h"
+
+namespace knowhere::sparse {
+template <typename T>
+class InvertedIndex {
+ public:
+    explicit InvertedIndex() {
+    }
+
+    void
+    SetUseWand(bool use_wand) {
+        std::unique_lock<std::shared_mutex> lock(mu_);
+        use_wand_ = use_wand;
+    }
+
+    Status
+    Save(MemoryIOWriter& writer) {
+        /**
+         * zero copy is not yet implemented, now serializing in a zero copy
+         * compatible way while still copying during deserialization.
+         *
+         * Layout:
+         *
+         * 1. int32_t rows, sign indicates whether to use wand
+         * 2. int32_t cols
+         * 3. for each row:
+         *     1. int32_t len
+         *     2. for each non-zero value:
+         *        1. table_t idx
+         *        2. T val
+         *     With zero copy deserization, each SparseRow object should
+         *     reference(not owning) the memory address of the first element.
+         *
+         * inverted_lut_ and max_in_dim_ not serialized, they will be
+         * constructed dynamically during deserialization.
+         *
+         * Data are densly packed in serialized bytes and no padding is added.
+         */
+        std::shared_lock<std::shared_mutex> lock(mu_);
+        writeBinaryPOD(writer, n_rows_internal() * (use_wand_ ? 1 : -1));
+        writeBinaryPOD(writer, n_cols_internal());
+        writeBinaryPOD(writer, value_threshold_);
+        for (size_t i = 0; i < n_rows_internal(); ++i) {
+            auto& row = raw_data_[i];
+            writeBinaryPOD(writer, row.size());
+            writer.write(row.data_ptr(), row.size() * (sizeof(table_t) + sizeof(T)));
+        }
+        return Status::success;
+    }
+
+    Status
+    Load(MemoryIOReader& reader) {
+        std::unique_lock<std::shared_mutex> lock(mu_);
+        int64_t rows;
+        readBinaryPOD(reader, rows);
+        use_wand_ = rows > 0;
+        rows = std::abs(rows);
+        size_t dim;
+        readBinaryPOD(reader, dim);
+        readBinaryPOD(reader, value_threshold_);
+
+        raw_data_.resize(rows);
+        inverted_lut_.resize(dim);
+        if (use_wand_) {
+            max_in_dim_.resize(dim);
+        }
+
+        for (int64_t i = 0; i < rows; ++i) {
+            auto& row = raw_data_[i];
+            size_t count;
+            readBinaryPOD(reader, count);
+            row.reserve(count);
+            reader.read(row.data_ptr(), count * (sizeof(table_t) + sizeof(T)));
+            add_row_to_index(row, i);
+        }
+
+        return Status::success;
+    }
+
+    // Non zero drop ratio is only supported for static index, i.e. data should
+    // include all rows that'll be added to the index.
+    Status
+    Train(const SparseRow<T>* data, size_t rows, float drop_ratio_build) {
+        if (drop_ratio_build == 0.0f) {
+            return Status::success;
+        }
+
+        std::vector<T> vals;
```

Comment:
just compute the size that needs to be preallocated, such as
```
size_t amount = 0;
for (size_t i = 0; i < rows; i++) { amount += data[i].size(); }

std::vector<T> vals(amount);
```

Reply:
Done